README.md
# react-leaflet-cluster-layer

`react-leaflet-cluster-layer` provides a simple `<ClusterLayer />` component for plotting React components as markers and clusters in a `react-leaflet` map.

![A screenshot of a cluster on a leaflet map](../master/screenshot.jpg?raw=true)

## Usage

```js
import React from 'react';
import { render } from 'react-dom';
import { Map, Marker, Popup, TileLayer } from 'react-leaflet';
import ClusterLayer from 'react-leaflet-cluster-layer';

const position = { lng: -122.673447, lat: 45.522558 };
const markers = [
  {
    position: { lng: -122.673447, lat: 45.5225581 },
    text: 'Voodoo Doughnut',
  },
  {
    position: { lng: -122.6781446, lat: 45.5225512 },
    text: 'Bailey\'s Taproom',
  },
  {
    position: { lng: -122.67535700000002, lat: 45.5192743 },
    text: 'Barista'
  }
];

class ExampleClusterComponent extends React.Component {

  render() {
    const style = {
      border: 'solid 2px darkgrey',
      borderRadius: '8px',
      backgroundColor: 'white',
      padding: '1em',
      textAlign: 'center'
    };
    const cluster = this.props.cluster;

    if (cluster.markers.length == 1) {
      return (
        <div style={style} >{cluster.markers[0].text}</div>
      );
    }

    return (
      <div style={style}>{cluster.markers.length} items</div>
    );
  }

}

const map = (
  <Map center={position} zoom={13}>
    <ClusterLayer
      markers={markers}
      clusterComponent={ExampleClusterComponent} />
    <TileLayer
      url='http://{s}.tile.osm.org/{z}/{x}/{y}.png'
      attribution='&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors'
    />
  </Map>
);

render(map, document.getElementById('app'));
```

## API

The `ClusterLayer` component takes the following props:

- `markers`: an array of objects that expose the properties defined in the `Marker` type
- `clusterComponent`: (required) the React component to be rendered for each marker and cluster, this component will receive the following props
  - `cluster`: a `Cluster` object, as defined by the Cluster Flow type
  - `style`: a style object for positioning
  - `map`: the Leaflet map object from the `react-leaflet` `MapLayer`
  - `...propsForClusters`: the component will also receive the properties of `propsForClusters` as props
- `propsForClusters`: props to pass on to marker and cluster components
- `gridSize`: optional prop to control how bounds of clusters expand while being generated (default: 60)
- `minClusterSize`: optional prop to enforce a minimum cluster size (default: 2)

## Example

To try the example:

1. Clone this repository
2. run `npm install` in the root of your cloned repository
3. run `npm run example`
4. Visit [localhost:8000](http://localhost:8000)

## Contributing

See [CONTRIBUTING.md](https://www.github.com/OpenGov/react-leaflet-cluster-layer/blob/master/CONTRIBUTING.md)

## License

`react-leaflet-cluster-layer` is MIT licensed.

See [LICENSE.md](https://www.github.com/OpenGov/react-leaflet-cluster-layer/blob/master/LICENSE.md) for details.


CHANGELOG.md
# 0.0.4 Release

- Lib is compatible with React 15
- Lib is compatible with react-leaflet 1.1

# 0.0.3 Release

- Correct issue introduced with removal of 'moveend' listener

# 0.0.2 Release

- Simplify event responders
- Fix a bug where clusters got wrong props

# 0.0.1 Release

- Initial implementation of package
- Updated readme, license filename, added changelog, added contribution guide
- Added description of technology and language to contribution guide
- Expanded instructions for running example in README.md
- Added unit tests
- Added developer certificate of origin to contributing guide


src/__tests__/ClusterLayer.test.js
import React from 'react';
import { shallow, mount, render } from 'enzyme';
import { Map } from 'react-leaflet';
import ClusterLayer from '../ClusterLayer.js';

jest.unmock('../ClusterLayer.js');

const L = jest.genMockFromModule('leaflet');

class ClusterComponent extends React.Component {
  render() {
    return (
      <div className="cluster-component">Cluster component</div>
    );
  }
}

describe('ClusterLayer', () => {

  /*
    Here we declare mocks or fixtures of
    all the data structures that the component uses
  */
  const center = { lng: -122.673447, lat: 45.522558 };

  const mockBounds = {
    contains: jest.fn(() => true),
    extend: jest.fn(),
    getNorthEast: jest.fn(() => ({ lng: -122, lat: 46 })),
    getSouthWest: jest.fn(() => ({ lng: -123, lat: 45 })),
  };

  const mockPanes = { overlayPane: { appendChild: jest.fn() } };

  const mockMap = {
    layerPointToLatLng: jest.fn(() => ({ lng: -122.6, lat: 45.522 })),
    latLngToLayerPoint: jest.fn(() => ({ x: 100, y: 100 })),
    on:  jest.fn(),
    getBounds:  jest.fn(() => mockBounds),
    getPanes: jest.fn(() => mockPanes),
    invalidateSize: jest.fn()
  };

  const mockMarkers = [
    {
      position: { lng: -122.673447, lat: 45.5225581 },
      text: 'Voodoo Doughnut',
    },
    {
      position: { lng: -122.6781446, lat: 45.5225512 },
      text: 'Bailey\'s Taproom',
    },
    {
      position: { lng: -122.67535700000002, lat: 45.5192743 },
      text: 'Barista'
    }
  ];

  it('should render', () => {
    const layer = render(
        <ClusterLayer
          markers={[]}
          clusterComponent={ClusterComponent} />
    );
    expect(layer).toBeTruthy();
  });

  it('should render a single child <ClusterComponent /> given one marker', () => {
    const layer = mount(
        <ClusterLayer
          map={mockMap}
          markers={[mockMarkers[0]]}
          clusterComponent={ClusterComponent} />
    );
    expect(layer.find('.cluster-component').length).toEqual(1);
  });

  it('should render one child <ClusterComponent /> given three markers with the same position', () => {
    const layer = mount(
        <ClusterLayer
          map={mockMap}
          markers={[mockMarkers[0], mockMarkers[0], mockMarkers[0]]}
          clusterComponent={ClusterComponent} />
    );
    expect(layer.find('.cluster-component').length).toEqual(1);
  });

  it('should render three child <ClusterComponent /> given three markers with the different positions', () => {
    const layer = mount(
        <ClusterLayer
          map={mockMap}
          markers={mockMarkers}
          clusterComponent={ClusterComponent} />
    );
    expect(layer.find('.cluster-component').length).toEqual(3);
  });

  it('should pass the `propsForClusters` prop to rendered <ClusterComponent />', () => {
    const mockProps = {
      theAnswer: 42,
      numCoffees: 3
    };

    const layer = mount(
        <ClusterLayer
          map={mockMap}
          markers={mockMarkers}
          propsForClusters={mockProps}
          clusterComponent={ClusterComponent} />
    );
    const componentProps = layer.find(ClusterComponent).at(0).props();
    expect(componentProps).toBeTruthy();
    expect(componentProps.theAnswer).toEqual(mockProps.theAnswer);
    expect(componentProps.numCoffees).toEqual(mockProps.numCoffees);
  });
});


src/ClusterLayer.js
import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';

import L from 'leaflet';
import {MapLayer} from 'react-leaflet';

export type LngLat = {
  lng: number;
  lat: number;
}

export type Marker = {
  position: LngLat;
  isAdded: boolean;
}

export type Point = {
  x: number;
  y: number;
}

export type Bounds = {
  contains: (latLng: LngLat) => boolean;
  extend: (latLng: LngLat) => void;
  getNorthEast: () => LngLat;
  getSouthWest: () => LngLat;
}

export type Map = {
  layerPointToLatLng: (lngLat: Point) => LngLat;
  latLngToLayerPoint: (lngLat: LngLat) => Point;
  on: (event: string, handler: () => void) => void;
  getBounds: () => Bounds;
  getPanes: () => Panes;
  invalidateSize: () => void;
}

export type Panes = {
  overlayPane: Pane;
}

export type Pane = {
  appendChild: (element: Object) => void;
}

export type Cluster = {
  center: LngLat;
  markers: Array<Marker>;
  bounds: Bounds;
}

// Taken from http://stackoverflow.com/questions/1538681/how-to-call-fromlatlngtodivpixel-in-google-maps-api-v3/12026134#12026134
// and modified to use Leaflet API
function getExtendedBounds(map: Map, bounds: Bounds, gridSize: number): Bounds {
  // Turn the bounds into latlng.
  const northEastLat = bounds && bounds.getNorthEast() && bounds.getNorthEast().lat;
  const northEastLng = bounds && bounds.getNorthEast() && bounds.getNorthEast().lng;
  const southWestLat = bounds && bounds.getSouthWest() && bounds.getSouthWest().lat;
  const southWestLng = bounds && bounds.getSouthWest() && bounds.getSouthWest().lng;

  const tr = L.latLng(northEastLat, northEastLng);
  const bl = L.latLng(southWestLat, southWestLng);

  // Convert the points to pixels and the extend out by the grid size.
  const trPix = map.latLngToLayerPoint(tr);
  trPix.x += gridSize;
  trPix.y -= gridSize;

  const blPix = map.latLngToLayerPoint(bl);
  blPix.x -= gridSize;
  blPix.y += gridSize;

  // Convert the pixel points back to LatLng
  const ne = map.layerPointToLatLng(trPix);
  const sw = map.layerPointToLatLng(blPix);

  // Extend the bounds to contain the new bounds.
  bounds.extend(ne);
  bounds.extend(sw);

  return bounds;
}

function distanceBetweenPoints(p1: LngLat, p2: LngLat): number {
  if (!p1 || !p2) {
    return 0;
  }

  const R = 6371; // Radius of the Earth in km

  const degreesToRadians = degree => (degree * Math.PI / 180);
  const sinDouble = degree => Math.pow(Math.sin(degree / 2), 2);
  const cosSquared = (point1, point2) => {
    return Math.cos(degreesToRadians(point1.lat)) * Math.cos(degreesToRadians(point2.lat));
  };

  const dLat = degreesToRadians(p2.lat - p1.lat);
  const dLon = degreesToRadians(p2.lng - p1.lng);
  const a = sinDouble(dLat) + cosSquared(p1, p2) * sinDouble(dLon);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  const d = R * c;
  return d;
}

export default class ClusterLayer extends MapLayer {
  static propTypes = {
    markers: PropTypes.array,
    clusterComponent: PropTypes.func.isRequired,
    propsForClusters: PropTypes.object,
    gridSize: PropTypes.number,
    minClusterSize: PropTypes.number
  };

  state: Object = {
    clusters: []
  };

  map: Object = null;

  constructor(props, context) {
    super(props, context);
    this.map = (undefined !== props.map) ? props.map : context.map;
  }

  createLeafletElement(props: Object): Object {
    return null
  }

  componentDidMount(): void {
    this.leafletElement = ReactDOM.findDOMNode(this.refs.container);
    this.map.getPanes().overlayPane.appendChild(this.leafletElement);
    this.setClustersWith(this.props.markers);
    this.attachEvents();
  }

  componentWillReceiveProps(nextProps: Object): void {
    if (!this.props || nextProps.markers !== this.props.markers) {
      this.setClustersWith(nextProps.markers);
    }
  }

  componentWillUnmount(): void {
    this.map.getPanes().overlayPane.removeChild(this.leafletElement);
  }

  componentDidUpdate(): void {
    this.map.invalidateSize();
    this.updatePosition();
  }

  shouldComponentUpdate(): boolean {
    return true;
  }

  setClustersWith(markers: Array<Marker>): void {
    this.setState({
      clusters: this.createClustersFor(markers)
    });
  }

  recalculate(): void {
    this.setClustersWith(this.props.markers);
    this.updatePosition();
  }

  attachEvents(): void {
    const map: Map = this.map;

    map.on('viewreset', () => this.recalculate());
    map.on('moveend', () => this.recalculate());
  }

  updatePosition(): void {
    this.state.clusters.forEach((cluster: Cluster, i) => {
      const clusterElement = ReactDOM.findDOMNode(
        this.refs[this.getClusterRefName(i)]
      );

      L.DomUtil.setPosition(
        clusterElement,
        this.map.latLngToLayerPoint(cluster.center)
      );
    });
  }

  render(): React.Element {
    return (
      <div ref="container"
         className={`leaflet-objects-pane
       leaflet-marker-pane
       leaflet-zoom-hide
       react-leaflet-cluster-layer`}
      >
        {this.renderClusters()}
      </div>
    );
  }

  renderClusters(): Array<React.Element> {
    const style = {position: 'absolute'};
    const ClusterComponent = this.props.clusterComponent;
    return this.state.clusters
      .map((cluster: Cluster, index: number) => (
        <ClusterComponent
          {...this.props.propsForClusters}
          key={index}
          style={style}
          map={this.map}
          ref={this.getClusterRefName(index)}
          cluster={cluster}
        />
      ));
  }

  getGridSize(): number {
    return this.props.gridSize || 60;
  }

  getMinClusterSize(): number {
    return this.props.minClusterSize || 2;
  }

  isMarkerInBounds(marker: Marker, bounds: Bounds): boolean {
    return bounds.contains(marker.position);
  }

  calculateClusterBounds(cluster: Cluster) {
    const bounds = L.latLngBounds(cluster.center, cluster.center);
    cluster.bounds = getExtendedBounds(this.map, bounds, this.getGridSize());
  }

  isMarkerInClusterBounds(cluster: Cluster, marker: Marker): boolean {
    return cluster.bounds.contains(L.latLng(marker.position));
  }

  addMarkerToCluster(cluster: Cluster, marker: Marker) {
    const center = cluster.center;
    const markersLen = cluster.markers.length;

    if (!center) {
      cluster.center = L.latLng(marker.position);
      this.calculateClusterBounds(cluster);
    } else {
      const len = markersLen + 1;
      const lng = (center.lng * (len - 1) + marker.position.lng) / len;
      const lat = (center.lat * (len - 1) + marker.position.lat) / len;
      cluster.center = L.latLng({lng, lat});
      this.calculateClusterBounds(cluster);
    }

    marker.isAdded = true;
    cluster.markers.push(marker);
  }

  createClustersFor(markers: Array<Marker>): Array<Cluster> {
    const map: Map = this.map;
    const extendedBounds = getExtendedBounds(map, map.getBounds(), this.getGridSize());
    return markers
      .filter(marker => extendedBounds.contains(L.latLng(marker.position)))
      .reduce((clusters: Array<Cluster>, marker: Marker) => {
        let distance = 40000; // Some large number
        let clusterToAddTo = null;
        const pos = marker.position;

        clusters.forEach((cluster) => {
          const center = cluster.center;
          if (center) {
            const d = distanceBetweenPoints(center, pos);
            if (d < distance) {
              distance = d;
              clusterToAddTo = cluster;
            }
          }
        });

        if (clusterToAddTo && this.isMarkerInClusterBounds(clusterToAddTo, marker)) {
          this.addMarkerToCluster(clusterToAddTo, marker);
        } else {
          const cluster = {
            markers: [marker],
            center: L.latLng(pos),
            bounds: L.latLngBounds()
          };
          this.calculateClusterBounds(cluster);
          clusters.push(cluster);
        }
        return clusters;
      }, []);
  }

  getClusterRefName(index: number): string {
    return `cluster${index}`;
  }

}


example/webpack.config.js
/* eslint-disable */
var webpack = require('webpack');

module.exports = {
  devtool: 'source-map',
  entry: {
    app: __dirname + '/index.js'
  },
  module: {
    loaders: [
      {
        test: /\.js$/,
        exclude: /node_modules/,
        loader: 'babel-loader',
        query: {
          plugins: [
            ['react-transform', {
              transforms: [{
                transform: 'react-transform-hmr',
                imports: ['react'],
                locals: ['module']
              }]
            }]
          ]
        }
      },
    ]
  },
  output: {
    path: __dirname + '/build/',
    filename: '[name].js',
    publicPath: 'http://localhost:8000/build'
  },
  plugins: [
    new webpack.DefinePlugin({
      'process.env': {
        'NODE_ENV': '"development"'
      }
    }),
    new webpack.NoEmitOnErrorsPlugin(),
    new webpack.HotModuleReplacementPlugin()
  ],
  devServer: {
    contentBase: __dirname,
    historyApiFallback: true,
    hot: true,
    inline: true,
    port: 8000,
    stats: {
      cached: false
    }
  }
};


example/index.js
import React from 'react';
import { render } from 'react-dom';
import { Map, Marker, Popup, TileLayer } from 'react-leaflet';
import ClusterLayer from '../src/ClusterLayer';

const position = { lng: -122.673447, lat: 45.522558 };
const markers = [
  {
    position: { lng: -122.673447, lat: 45.5225581 },
    text: 'Voodoo Doughnut',
  },
  {
    position: { lng: -122.6781446, lat: 45.5225512 },
    text: 'Bailey\'s Taproom',
  },
  {
    position: { lng: -122.67535700000002, lat: 45.5192743 },
    text: 'Barista'
  },
  {
    position: { lng: -122.65596570000001, lat: 45.5199148000001 },
    text: 'Base Camp Brewing'
  }
];

class ExampleClusterComponent extends React.Component {

  render() {
    const style = {
      border: 'solid 3px lightblue',
      backgroundColor: '#333333',
      color: 'white',
      textAlign: 'center',
      margin: '0',
      padding: '0.25em 0.25em 0.5em',
      fontSize: '1.25em',
      fontWeight: '700'
    };
    const cluster = this.props.cluster;

    if (cluster.markers.length == 1) {
      const markerStyle = Object.assign({}, style, {
        minWidth: '110px',
        borderRadius: '16px',
        borderTopLeftRadius: '0',
      });

      return (
        <div style={markerStyle} >{cluster.markers[0].text}</div>
      );
    }

    const clusterStyle = Object.assign({}, style, {
      borderRadius: '50%',
      borderTopLeftRadius: '0',
      width: '24px',
      height: '24px'
    });

    return (
      <div style={clusterStyle}>{cluster.markers.length}</div>
    );
  }

}

const map = (
  <Map center={position} zoom={13}>
    <ClusterLayer
      markers={markers}
      clusterComponent={ExampleClusterComponent} />
    <TileLayer
      url='http://{s}.tile.osm.org/{z}/{x}/{y}.png'
      attribution='&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors'
    />
  </Map>
);

render(map, document.getElementById('app'));
